src/PureScript/CST/ModuleGraph.purs

```diff
@@ -0,0 +1,117 @@
+module PureScript.CST.ModuleGraph
+  ( moduleGraph
+  , sortModules
+  , ModuleSort(..)
+  ) where
+
+import Prelude
+
+import Data.Array as Array
+import Data.Bifunctor (bimap)
+import Data.Either (Either(..))
+import Data.Foldable (all, foldl)
+import Data.Functor.Compose (Compose(..))
+import Data.Identity (Identity(..))
+import Data.List (List(..))
+import Data.List as List
+import Data.Map (Map)
+import Data.Map as Map
+import Data.Maybe (Maybe(..), fromMaybe, isJust, maybe)
+import Data.Newtype (un)
+import Data.Set (Set)
+import Data.Set as Set
+import Data.Tuple (Tuple(..))
+import Control.Monad.Free (Free, runFree)
+import PureScript.CST.Types (ImportDecl(..), ModuleHeader(..), ModuleName, Name(..))
+
+type Graph a = Map a (Set a)
+
+moduleGraph :: forall e. Array (ModuleHeader e) -> Graph ModuleName
+moduleGraph = Map.fromFoldable <<< map go
+  where
+  go (ModuleHeader { name: Name { name }, imports }) =
+    Tuple name (Set.fromFoldable (map getImportName imports))
+
+  getImportName (ImportDecl { "module": Name { name } }) = name
+
+data ModuleSort e
+  = Sorted (Array (ModuleHeader e))
+  | CycleDetected (Array (ModuleHeader e))
+
+sortModules :: forall e. Array (ModuleHeader e) -> ModuleSort e
+sortModules moduleHeaders = do
+  let
+    knownModuleHeaders :: Map ModuleName (ModuleHeader e)
+    knownModuleHeaders =
+      moduleHeaders
+        # map (\header@(ModuleHeader { name: Name { name } }) -> Tuple name header)
+        # Map.fromFoldable
+
+    graph = moduleGraph moduleHeaders
+    lookupModuleHeaders = Array.mapMaybe (flip Map.lookup knownModuleHeaders) <<< List.toUnfoldable
+
+  case bimap lookupModuleHeaders lookupModuleHeaders (topoSort graph) of
+    Left cycle -> CycleDetected cycle
+    Right sorted -> Sorted sorted
+
+type TopoSortArgs a =
+  { roots :: Set a
+  , sorted :: List a
+  , usages :: Map a Int
+  }
+
+topoSort :: forall a. Ord a => Graph a -> Either (List a) (List a)
+topoSort graph = do
+  _.sorted <$> go { roots: startingModules, sorted: Nil, usages: importCounts }
+  where
+  go :: TopoSortArgs a -> Either (List a) (TopoSortArgs a)
+  go { roots, sorted, usages } = case Set.findMin roots of
+    Nothing ->
+      if all (eq 0) usages then
+        Right { roots, sorted, usages }
+      else do
+        let
+          nonLeaf = Set.fromFoldable $ Map.keys $ Map.filterWithKey (\a count -> count /= 0 && Map.lookup a graph /= Nothing && Map.lookup a graph /= Just Set.empty) usages
+
+        case foldl (\b a -> if isJust b then b else runFree (un Identity) (un Compose (depthFirst { path: Nil, visited: Set.empty, curr: a }))) Nothing nonLeaf of
```

Comment:
Can we do something to cleanup the horizontal length of this case head?

Reply:
Yeah, removing the `Free` stuff helps and I'll move it to the let binding.